fix flaky timelion expression typeahead test

nreese opened the pull request:
fixes https://github.com/elastic/kibana/issues/17407

 The functional test flakiness results from 3 UI bugs.
1) There is not loading indicator when dynamic suggestions are getting fetched via async calls. This is the source of the test flakiness. The functional tests (just like the user) do not know when an async call is finished. The functional tests work most of the time when the results are returned fast enough. They break when the results are returned consistently slowly.
2) The old suggestions stay displayed until the async call returns
3) There is no logic to ensure that out of order responses are properly handled.

To get the test to fail consistently, add the below lines [here](https://github.com/elastic/kibana/blob/master/src/core_plugins/timelion/public/directives/timelion_expression_suggestions/arg_value_suggestions.js#L158). This will cause the async calls to be consistently slow so the test fails all the time.
```
      function timeout() {
        return new Promise(resolve => setTimeout(resolve, 3000));
      }
      await timeout();
```

This PR fixes the UI problems by adding a loading indicator to the UI, clear the old suggestions when new suggestions are requested, and properly handling out of order responses. The PR also revamps the functional tests so they are easier to follow.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7865/)

nreese commented:
no flakyness - try again

jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7888/)

nreese commented:
got a test failure but not the target timelion test failure

```
fail: "visualize app input control visualization updateFiltersOnChange is false should replace existing filter pill(s) when new item is selected"
20:39:33    │ proc  [ftr]        │        Error: expected true to equal false
```

jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7916/)

timroes commented:
Jenkins, test this - flaky test

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7991/)

timroes commented:
Jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8486/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8850/)

nreese commented:
Dynamic suggestions are ones that are pulled via an async call versus the hard coded suggestions. The idea was to provide type ahead support for knowledge that is in saved objects but is a pain to remember. For example, when you are setting the index, why not just display a list of index-patterns, when setting the timefield, why not show all date fields if the index matches an index-pattern and so on. I put this in because I could never remember the names of index-patterns for field names and it drove me crazy to have to open another window and look and the index pattern management page to see what was available.

To view dynamic suggestions in action, here a few example expressions. Just position your cursor after the `=` char.
1) `.es(index=)`
2) `.es(index=logstash-*,timefield=)`
3) `.es(index=logstash-*,split=)`

Stacey-Gammon commented on src/core_plugins/timelion/public/directives/timelion_expression_input.js:
```diff
@@ -158,8 +162,14 @@ app.directive('timelionExpressionInput', function ($document, $http, $interval,
           argValueSuggestions
         );
 
+        if (_.get(suggestions, 'isPrevRequestResults', false)) {
```
I don't think this is the right way to do this (it also doesn't seem to be working well in practice - I'll post a gif shortly)

I think instead what we should do is return the input the suggestions are based off of on the suggestions object.  Then when we get here, do something like  `if (suggestions.sheet != scope.sheet) return;` along with a comment that says something like, "our sheet changed since we requested the suggestions, they are no longer valid."

I'm only saying "sheet" because that's what's on scope it seems like.

nreese replied:
Good idea. This is a lot simpler and easier to follow

Stacey-Gammon commented:
This doesn't seem to be working well (it's hard to see but first I type 'a', then delete it and type 'k' so I get back three sets of responses in a different order = empty string, a, then k.

![wrongsuggestions](https://user-images.githubusercontent.com/16563603/43731416-711b8f34-997c-11e8-92b3-9744dfb516d6.gif)

see comment for an alternative way that I think improves the code and hopefully will fix this. I experimented with the different return times by using your suggestion above but adding a random number:

```
      const random = Math.floor(Math.random() * Math.floor(10));
      console.log('getDynamicSuggestionsForArgument for ' + partialInput + ' will return after ' + random + ' seconds');
      function timeout() {
        return new Promise(resolve => setTimeout(resolve, random * 1000));
      }
      await timeout();
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8902/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8912/)

Stacey-Gammon commented on src/core_plugins/timelion/public/directives/timelion_expression_input.js:
```diff
@@ -150,16 +150,26 @@ app.directive('timelionExpressionInput', function ($document, $http, $interval,
+
         const suggestions = await suggest(
-          scope.sheet,
+          scope.sheet,  // sheet is the timelion expression
```
I just had a thought... I wonder if you could avoid saving the expression on the returned expressions if you just saved a local variable here.  
```
const expression = scope.sheet; // maybe need to make a copy??
const suggestions = await suggest(expression...);
if (scope.sheet !== expression) {
  // Expression changed since requesting, no longer valid
  return;
}
```

Could then lessen the footprint of this PR even more.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8972/)

Stacey-Gammon commented:
When I added the code in to change up the returned objects, I hit a console error and got no suggestions:

```
angular.js:14800 TypeError: Cannot read property 'sort' of undefined
    at Suggestions.setList (timelion_expression_input_helpers.js:308)
    at timelion_expression_input.js:120
    at Scope.$eval (angular.js:18542)
    at Scope.$apply (angular.js:18641)
    at _callee$ (timelion_expression_input.js:117)
    at tryCatch (runtime.js:65)
    at Generator.invoke [as _invoke] (runtime.js:303)
    at Generator.prototype.(anonymous function) [as next] (http://localhost:5601/otm/bundles/vendors.bundle.js:51975:21)
    at step (timelion_expression_input.js:33)
    at timelion_expression_input.js:33
```

<img width="593" alt="screen shot 2018-08-07 at 12 55 58 pm" src="https://user-images.githubusercontent.com/16563603/43790679-4fad3a8a-9a41-11e8-933f-c24a5dd6629b.png">

Not exactly sure where the issue is. If it's legit, think you could simulate this error, where responses show up in a different order, in a test?

nreese commented:
> When I added the code in to change up the returned objects, I hit a console error and got no suggestions:

I could not reproduce with the latest. I added the random code to change the order and never ran into any problems

Stacey-Gammon commented:
Darnit, now I can't either.  Oh well, hopefully it was a local issue.

Only last thing I notice is that there is a flicker when the response is really fast.  Not the end of the world. I thought there was a standard way to get rid of that flicker.  Like add in a debounce or something so if the response comes back in < 200ms then no "Loading" indicator will be shown.

cc @cjcenizal - does this kind of thing sound familiar to you?

timroes commented:
@markov00 @ppisljar Could one of you perhaps giving this PR a review?
